### PR TITLE
add dashboard module

### PR DESCRIPTION
**Pull request by ulferts**
Adds dashboard functionality to OP allowing a user to configure a page for e.g. reporting purposes.

This PR can be merged once a minimal viable set of widgets can be configured or, if that cannot be reached in sufficient time and it needs to be merged for other reasons, if the functionality can be hidden.

To keep things simple, only a single page is currently supported. Subpages can be added later.

https://community.openproject.com/projects/openproject/work_packages/30422

# TODO
* [x] setup dashboard ruby module
* [x] Add dashboard angular module displaying a grid
* [x] Extend grid backend to allow dashboard pages
* [x] Add widgets. They can probably be converted from the my page in parts.
  * [x] work packages table (scoped to project)
  * [x] work packages graph
  * [x] clarify if further widgets are required for a minimal set
* [x] Have an icon for the dashboards menu entry
* [x] Cleanup duplicate code between wp-table and wp-graph
* [x] Migrate wp-assigned et. al. to wp-table

# Caveats
* As all table widgets now create a work package query, a user needs to have the `save_queries` permission to add such widgets. This might decrease the availability of such widgets on the my page. We should think about having more global permissions where `save_queries` is enabled for all users by default.

### CODE REVIEW

File: db/migrate/20190710132957_rename_my_page_widgets.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
+class RenameMyPageWidgets < ActiveRecord::Migration[5.2]
+  def up
+    Grids::MyPage.includes(:widgets, user: :preference).each do |page|
+      I18n.with_locale(page.user.language || 'en') do
```
Comment (oliverguenther) [inline db/migrate/20190710132957_rename_my_page_widgets.rb]:
```suggestion
      I18n.with_locale(page.user&.language.presence || 'en') do
```

On my local community dump, users partially have an empty string as a language, or the `page.user` is nil, both times breaking the migration.

**Comment by oliverguenther** [comment]
It is difficult to add new rows to the dashboard due to the overlay to resize the main menu. Also the widget was added with a single row by default which does not leave enough space. In this case we should probably auto-add a row because it's not possible to resize the widget until a row was added

Peek 2019-07-11 14-26

**Comment by oliverguenther** [comment]
Looks good to me behavior wise. I've noticed that all new components use `ChangeDetectionStrategy.default`. Unfortunately it's currently not possible to use `onPush` as the default, so you have to manually set this change detection manually. I would suggest we do that for the dashboard grid and all subcomponents.

Ideally we would set this also in the `ApplicationBase` component, which then forces all children to become OnPush implicitly. This is breaking the my page and some other components still, but in preparation for this, please convert the dashboard grid components to OnPush manually. It's fine for me to do that in separate PRs however. Feel free to merge in that case.

File: db/migrate/20190710132957_rename_my_page_widgets.rb (hunk 1 above)
Comment (ulferts) [reply db/migrate/20190710132957_rename_my_page_widgets.rb]: thanks

**Comment by ulferts** [comment]
@oliverguenther I will work on transforming the widgets one by one with subsequent PRs. Regarding the layout shortcomings we have a separate ticket where we will evaluate different layouts. Feel free to chime in: https://community.openproject.com/projects/openproject/work_packages/30544

**Pull request merged**